README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Overview

The R package NearRepeat uses the Knox test for space-time clustering to
quantify the spatio-temporal association between events. In criminology,
this test has been used to identify people and locations that are at
disproportionate risk of victimization. Of interest is often not only
the ‘same repeat’ victims (people or locations that after a first crime
event, are targeted again within a short period thereafter), but also
the ‘near repeat’ victims: **nearby** people or locations that are
victimized within a short period after the first crime. For more
information, see for example the [JDI Analysis
Brief](http://www.ucl.ac.uk/jdibrief/analysis/repeat_victimisation).

Note that in parallel with the development of the code here, Toby Davies
has developed an alternative implementation in Python, which can be
found [here](https://github.com/tobydavies/NearRepeat). We have
collaborated on some of the accompanying materials and reviewed each
other’s code, and as far as we are aware the two implementations should
give identical results.

Several other options exist but have drawbacks:

  - [Prof. Jerry Ratcliffe’s](http://www.jratcliffe.net/software/) /
    [Temple
    University’s](http://www.cla.temple.edu/center-for-security-and-crime-science/projects/#near-repeat-calculator)
    Near Repeat calculator only works on Windows (and needs admin rights
    to execute and write files). It seems this software mislabels the
    rows and columns of its output files, which may lead to erroneous
    conclusions.

  - the free [JDI Near Repeat
    Toolkit](http://www.ucl.ac.uk/jill-dando-institute/research/research-groups/geo-crime/near_repeat_toolkit)
    also only works on Windows. However it is not actively supported and
    does not run on Windows 10.

  - the R package
    [surveillance](https://cran.r-project.org/package=surveillance) has
    implemented the `knox()` function, but this only works for a single
    spatial distance and temporal distance, whereas the current package
    also works with multiple bands of spatial and temporal distances, as
    well as same repeat analysis.

## Version

The most recent version of the package is:

> Steenbeek, W. (2018). *Near Repeat*. R package version 0.1.1. URL:
> <https://github.com/wsteenbeek/NearRepeat>

## Installation

You can install the package from this [GitHub
repository](https://github.com/wsteenbeek/NearRepeat). You first need to
install the [remotes](https://CRAN.R-project.org/package=remotes)
package.

``` r
install.packages("remotes")
```

Then install the `NearRepeat` package using the `install_github`
function in the [remotes](https://CRAN.R-project.org/package=remotes)
package.

``` r
remotes::install_github("wsteenbeek/NearRepeat")
```

Note that if you also want to view the vignettes directly within R, the
command should read:

``` r
remotes::install_github("wsteenbeek/NearRepeat", build_vignettes = TRUE)
```

After a few moments you should see “creating vignettes …” and at that
point you have to wait a few minutes. Because of the wait time and
because you may need additional packages to build vignettes, the default
`install_github()` call does not build vignettes. Alternatively, the
vignettes can also be viewed directly online (see section ‘Vignettes’).

Package `NearRepeat` needs packages `future.apply`, `progressr`, and
`ggplot2` to function correctly. If these are not installed already,
following the steps above should install these dependencies as well. If
this doesn’t happen, it’s best to run
`install.packages(c("future.apply", "progressr", "ggplot2"))` and try
the steps above again.

## Example

The package currently has one main function, also called `NearRepeat()`.
See the helpfile to see examples of the function in action:

``` r
library(NearRepeat)
?NearRepeat
```

## Vignettes

Vignettes are a long-form guide to an R package. Read these to get a
relatively easy introduction to the package and its various options. You
can view the available vignettes using:

``` r
browseVignettes("NearRepeat")
```

To directly read the vignettes rather than going through
`browseVignettes("NearRepeat")` you can use:

``` r
vignette("NearRepeat", package = "NearRepeat")
vignette("NearRepeat_breaks", package = "NearRepeat")
vignette("NRC", package = "NearRepeat")
vignette("prepare_data", package = "NearRepeat")
```

Finally, the vignettes can be viewed directly, without installing the
package, using the following links:

  - [NearRepeat: introduction to the
    package](http://htmlpreview.github.io/?https://github.com/wsteenbeek/NearRepeat/blob/master/doc/NearRepeat.html)
  - [NearRepeat\_breaks: a detailed description of how bandwidths can be
    specified](http://htmlpreview.github.io/?https://github.com/wsteenbeek/NearRepeat/blob/master/doc/NearRepeat_breaks.html)
  - [NRC: a comparison to the Near Repeat
    Calculator](http://htmlpreview.github.io/?https://github.com/wsteenbeek/NearRepeat/blob/master/doc/NRC.html)
  - [prepare\_data: how data was downloaded and prepped for the examples
    in the
    vignettes](http://htmlpreview.github.io/?https://github.com/wsteenbeek/NearRepeat/blob/master/doc/prepare_data.html)

## License

This package is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License, version 3, as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose. See the GNU General
Public License for more details.

A copy of the GNU General Public License, version 3, is available at
<https://www.r-project.org/Licenses/GPL-3>


NEWS.md
# NearRepeat 0.1.1

* Includes progress bar using `progressr` package

* Other minor cleanup of code

# NearRepeat 0.1.0

* First published version of the `NearRepeat` package.


R/plot_function.R
#' @import ggplot2
NULL

#' Plots a Heat Map of Near Repeat results based on p-values
#'
#' @param knox_object     a knox object, i.e. the result of a 'NearRepeat()' funtion call
#' @param pvalue_range    the range of p-values that will be colored as "significant",
#'                        default = c(0, .05)
#' @param text            labels of cells, default is "knox_ratio". Possible values are
#'                        "observed", "knox_ratio", "knox_ratio_median", "pvalues", or NA
#' @param minimum_perc    The minimum percentage of increased Knox value that will be highlighted.
#'                        Default is 20% (or a Knox ratio of >= 1.2), as per the recommendations
#'                        of Ratcliffe (2009). If parameter text equals "pvalues" or NA,
#'                        the p-values alone determine highlighting, irrespective of the Knox (median)
#'                        ratios.
#' @return                a heat map based on p-values (generated by ggplot2)
#' @examples
#'
#' # Generate example data. Suppose x and y refer to meters distance.
#' set.seed(10)
#' (mydata <- data.frame(x = sample(x = 20, size = 20, replace = TRUE) * 20,
#'                      y = sample(x = 20, size = 20, replace = TRUE) * 20,
#'                      date = as.Date(sort(sample(20, size = 20, replace = TRUE)),
#'                      origin = "2018-01-01")
#'                      ))
#'
#' # The plot() function can be used to plot a Heat Map of Near Repeat results based on p-values
#' set.seed(4622)
#' myoutput <- NearRepeat(x = mydata$x, y = mydata$y, time = mydata$date,
#'                        sds = c(0,100,200,300,400), td = c(0,1,2,3,4,5))
#' plot(myoutput)
#'
#' # The default range of p-values that will be highlighted (0-.05) can be adjusted using
#' # the 'pvalue_range' parameter. By default the Knox ratios are printed in the cells,
#' # but this can be adjusted using the 'text' parameter. The default is "knox_ratio".
#' # Possible values are "observed", "knox_ratio", "knox_ratio_median", "pvalues", or NA.
#' # For more information, see vignette("NearRepeat")
#'
#' plot(myoutput, pvalue_range = c(0, .1), text = "observed")
#' plot(myoutput, pvalue_range = c(0, .1), text = "pvalues")
#'
#' @export
plot.knox <- function(knox_object, pvalue_range = c(0, .05), text = "knox_ratio", minimum_perc = 20){

  ggplot.df <- data.frame(knox_object$pvalues)

  ggplot.df$observed <- as.vector(knox_object$observed)
  ggplot.df$knox_ratio <- as.vector(knox_object$knox_ratio)
  ggplot.df$knox_ratio_median <- as.vector(knox_object$knox_ratio_median)
  ggplot.df$pvalues <- ggplot.df$Freq

  if(!is.na(text)){
    if (text == "knox_ratio" | text == "knox_ratio_median"){
      ggplot.df$pvalues[ggplot.df[, text] < (100 + minimum_perc)/100] <- NA
    }
  }

  ggplot.df$Var1 <- factor(ggplot.df$Var1, levels = rev(levels(ggplot.df$Var1)))

  myplot <- ggplot(data = data.frame(ggplot.df), aes(x=Var2, y=Var1, fill=pvalues)) +
                   geom_tile(colour = "black") +
                   scale_fill_gradient2(low = "darkred", high = "khaki1", mid = "orangered",
                                        limit = pvalue_range, na.value = "white",
                                        name = "p-value") +
                   theme_minimal() +
                   theme(axis.text.x = element_text(angle = 45, hjust = .2, vjust = .9)) +
                   coord_fixed() +
                   xlab("Temporal distance") +
                   ylab("Spatial distance") +
                   scale_x_discrete(position = "top")

  if(!is.na(text)) myplot <- myplot + geom_text(aes(label = round(get(text), 2)))

  return(myplot)
}



R/NearRepeat.R
#' @import future.apply
#' @import progressr
NULL

#' Near Repeat calculator using the Knox test
#'
#' This function uses the Knox test for space-time clustering to quantify the spatio-temporal
#' association between events.
#'
#' (Please note that any cases with missing x, y, or time will be
#' removed, i.e. this crime event will not be included in the analysis.)
#'
#' Standard \href{https://en.wikipedia.org/wiki/Interval_(mathematics)}{interval notation}
#' is used to indicate whether break points of that interval are included or excluded from the interval.
#'
#' Use set.seed() before the function call to guarantee reproducibility. See helpfile and vignette 'NearRepeat' for details.
#'
#' @param x                  a vector of x coordinates
#' @param y                  a vector of y coordinates
#' @param time               a vector of time. This can be of type integer, numeric, or date
#' @param sds                A vector of break points of the spatial intervals. For example c(0,50,120,300) to specify
#'                           spatial intervals from 0-50, 50-120, 120-300 meters. Or c(0,50,100,Inf) to specify
#'                           spatial intervals from 0-50, 50-100, and 100-Inf meters. (More accurately,
#'                           on the scale of the provided x and y coordinates. For example, data may be
#'                           projected in feet and thus the distances refer to feet instead of meters).
#' @param tds                A vector of break points of the temporal intervals. For example c(0,2,4,Inf) to specify temporal intervals
#'                           from 0-2, 2-4, 4-Inf days.
#' @param s_include.lowest   the descriptions above are ambiguous on how exactly the spatial break points are handled. For example,
#'                           does c(0,100,200) refer to 0-100, 101-200? Or to 0-99 and 100-199?
#'                           s_include.lowest follows the arguments of cut (see ?cut). Logical, indicating if a spatial
#'                           distance equal to the lowest (or highest, for right = FALSE) ‘breaks’ value should be included.
#'                           Default = FALSE. See vignette("NearRepeat_breaks") for details.
#' @param s_right            logical, indicating if the spatial intervals should be closed on the right (and open on
#'                           the left) or vice versa. Default = FALSE. See vignette("NearRepeat_breaks") for details.
#' @param t_include.lowest   t_include.lowest follows the arguments of cut (see ?cut). Logical, indicating if a temporal
#'                           distance equal to the lowest (or highest, for right = FALSE) ‘breaks’ value should be included.
#'                           Default = FALSE. See vignette("NearRepeat_breaks") for details.
#' @param t_right            logical, indicating if the temporal intervals should be closed on the right (and open on
#'                           the left) or vice versa. Default = FALSE. See vignette("NearRepeat_breaks") for details.
#' @param method             The method to calculate the spatial distances between crime events. Methods possible
#'                           as in the 'dist' function (see ?dist). Default is 'manhattan', which seems to be a fair
#'                           approximation of the distance travelled by a road network. Alternatively, the user can
#'                           specify 'euclidean' to get the 'as the crow flies' distance.
#' @param nrep               The number of replications of the Monte Carlo simulation (default = 999).
#' @param saveSimulations    Should all simulated contingency tables be saved as a 3-dimensional array? Default = FALSE
#' @param future.seed        A logical or an integer (of length one or seven), or a list of length(X) with pre-generated
#'                           random seeds. Default = TRUE. See R package future.apply for details.
#' @param ...                (optional) Additional arguments passed to future_lapply()
#' @return                   An object of type "knox", i.e. a list with four tables. For each spatial and temporal distance combination,
#'                           (1) The counts of observed crime pairs, (2) The Knox ratios based on the mean of the
#'                           simulations, (3) The Knox ratios based on the median of the simulations, (4) p-values.
#' @examples
#'
#' # Generate example data. Suppose x and y refer to meters distance.
#' set.seed(10)
#' (mydata <- data.frame(x = sample(x = 20, size = 20, replace = TRUE) * 20,
#'                      y = sample(x = 20, size = 20, replace = TRUE) * 20,
#'                      date = as.Date(sort(sample(20, size = 20, replace = TRUE)),
#'                      origin = "2018-01-01")
#'                      ))
#'
#' # Near Repeat calculation using 0-100 meters and 100-Inf meters, and three
#' # temporal intervals of 2 days
#' set.seed(38673)
#' NearRepeat(x = mydata$x, y = mydata$y, time = mydata$date,
#'            sds = c(0,100,Inf), tds = c(0,2,4))
#'
#' # Add a 'same repeat' spatial interval of 0.001 meters, and use Euclidean
#' # distance
#' set.seed(38673)
#' NearRepeat(x = mydata$x, y = mydata$y, time = mydata$date,
#'            sds = c(0,0.001,100,Inf), tds = c(0,2,4),
#'            method = "euclidean")
#'
#' # Only do 99 replications
#' set.seed(38673)
#' NearRepeat(x = mydata$x, y = mydata$y, time = mydata$date,
#'            sds = c(0,0.001,100,Inf), tds = c(0,2,4),
#'            method = "euclidean", nrep = 99)
#'
#'
#' # The plot() function can be used to plot a Heat Map of Near Repeat results
#' # based on p-values
#' set.seed(4622)
#' myoutput <- NearRepeat(x = mydata$x, y = mydata$y, time = mydata$date,
#'                        sds = c(0,100,200,300,400), td = c(0,1,2,3,4,5))
#' plot(myoutput)
#'
#' # The default range of p-values that will be highlighted (0-.05) can be
#' # adjusted using the 'pvalue_range' parameter. By default the Knox ratios
#' # are printed in the cells, but this can be adjusted using the 'text'
#' # parameter. The default is "knox_ratio". Possible values are "observed",
#' # "knox_ratio", "knox_ratio_median", "pvalues", or NA.
#'
#' # For more information, see vignette("NearRepeat")
#'
#' plot(myoutput, pvalue_range = c(0, .1), text = "observed")
#' plot(myoutput, pvalue_range = c(0, .1), text = "pvalues")
#'
#' @export
NearRepeat <- function(x, y, time,
                       sds, tds,
                       s_include.lowest = FALSE, s_right = FALSE,
                       t_include.lowest = FALSE, t_right = FALSE,
                       method = "manhattan",
                       nrep = 999,
                       saveSimulations = FALSE,
                       future.seed = TRUE,
                       ...
){

  # check for at least 2 replications
  if (nrep < 2) stop("At least 2 replications are needed. Increase 'nrep'")

  # check for a vector of spatial and temporal intervals
  if (length(sds) < 2 | length(tds) < 2) stop("Supply a vector of spatial and temporal intervals")

  # Create data.frame
  mydf <- data.frame(x = x, y = y, time = time)

  # Missing data?
  if (sum(!stats::complete.cases(mydf)) > 0){
    # Print information on deleted cases
    print(paste0(sum(!stats::complete.cases(mydf))," events are removed due to missing x or y or time"))
    # Remove missing cases
    mydf <- mydf[stats::complete.cases(mydf),]
  }

  # Create xy matrix
  xy <- cbind(mydf$x, mydf$y)

  # Distances of the observed space-time pairs
  s_dist <- stats::dist(xy, method = method)
  t_dist <- stats::dist(mydf$time)

  # Observed space-time pairs
  observed <- table(cut(s_dist, sds, include.lowest = s_include.lowest, right = s_right, dig.lab = 10),
                    cut(t_dist, tds, include.lowest = t_include.lowest, right = t_right, dig.lab = 10))

  ########
  #
  # MC Simulation to compute the null distribution
  #
  ########

  # Spatial cuts done once here (to reuse many times below)
  spat_cut <- cut(s_dist, sds, include.lowest = s_include.lowest, right = s_right, dig.lab = 10)

  #### Run permutations, calculate pairs, and save to list
  progressr::handlers("progress")
  progressr::with_progress({
    p <- progressr::progressor(along = seq_len(nrep))
    mylist <- future_lapply(seq_len(nrep), function(i) {
      p()
      t_dist_perm <- stats::dist(sample(mydf$time))
      return(table(spat_cut, cut(t_dist_perm, tds, include.lowest = t_include.lowest,
                                 right = t_right, dig.lab = 10)))
    }, future.seed = future.seed, ...)
  })


  # Create 3-dimensional output array
  array_Knox <- array(data = as.numeric(unlist(mylist)), dim = c(dim(observed), nrep))

  ########
  #
  # Create output
  #
  ########

  knox_ratio <- observed / apply(array_Knox, 1:2, mean)
  knox_ratio_median <- observed / apply(array_Knox, 1:2, median)

  pvalues <- observed
  for (i in seq_len(nrow(pvalues))){
    for (j in seq_len(ncol(pvalues))){
      pvalues[i,j] <- (sum(array_Knox[i,j,] >= observed[i,j]) + 1) / (nrep + 1)
    }
  }

  # Create output list
  result <- list(observed = observed,
                 knox_ratio = knox_ratio,
                 knox_ratio_median = knox_ratio_median,
                 pvalues = pvalues,
                 array_Knox = array_Knox)
  # remove array_Knox if saveSimulations == FALSE
  if(!saveSimulations) result$array_Knox <- NULL

  class(result) <- "knox"

  return(result)
}
